Detect the correct set of properties to use for JHipster applications

bdemers opened the pull request:
Now that the CLI supports Quarkus, JHipster + Quarkus apps need to use the Quarkus property set.

Detection of quarkus is based on finding `generator-jhipster-quarkus` in the package.json in the current directory.

Fallback is Spring based properties

In the future, adding other property types (i.e. Micronaut can be done the same way)

bdemers commented on common/src/test/groovy/com/okta/cli/common/RestoreSystemProperties.java:
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2017-Present Okta, Inc.
```
(copied from elsewhere)

mraible commented:
@bdemers I tried this tonight with a JHipster + Quarkus app. It works for login but doesn't for logout. I'm not sure this is an issue with the CLI.

@danielpetisme With this PR, the Okta CLI creates an `.okta.env` file when you run `okta apps create jhipster`. It has data like the following in it:

```sh
export QUARKUS_OIDC_AUTH_SERVER_URL="https://dev-5597275.okta.com/oauth2/default"
export QUARKUS_OIDC_CLIENT_ID="0oa5hi4eiSkdnMcfL5d6"
export QUARKUS_OIDC_CREDENTIALS_SECRET="XXX"
export QUARKUS_OIDC_APPLICATION_TYPE="web-app"
export QUARKUS_OIDC_AUTHENTICATION_REDIRECT_PATH="/login/oauth2/code/oidc"
```

I'm able to log in with this config, but I'm unable to logout. It *looks* like I'm logged out, but if I click the "sign in" link again, it logs me in again without prompting for credentials. 

I tried the same app with Keycloak and logout works as expected. You're prompted to enter your credentials again after you logout.

bdemers commented:
@mraible I'm guessing the following props need to be set:

```properties
quarkus.oidc.logout.path=/logout
quarkus.oidc.logout.post-logout-path=/
```

Though I'm not sure what the _norm_ is for jhipster, If these match what jhipster expects we can set them in `OidcProperties.Quarkus`, let me know

mraible commented:
@bdemers I looked in my generated app's `application.properties` and noticed the logout URL property:

```properties
jhipster.oidc.logout-url=http://localhost:9080/auth/realms/jhipster/protocol/openid-connect/logout
```

If I add the following to my `.okta.env`:

```sh
export JHIPSTER_OIDC_LOGOUT_URL="${QUARKUS_OIDC_AUTH_SERVER_URL}/v1/logout"
```

It fixes the problem. Another thing I noticed is JHipster uses `hybrid` for the application type.

```properties
quarkus.oidc.application-type=hybrid
```

Is there any particular reason you're overriding this for Okta?

mraible commented on common/src/test/groovy/com/okta/cli/common/model/OidcPropertiesTest.groovy:
```diff
@@ -63,6 +67,95 @@ class OidcPropertiesTest {
+            "quarkus.oidc.auth-server-url": "https://issuer.example.com",
+            "quarkus.oidc.client-id": "test-client-id",
+            "quarkus.oidc.credentials.secret": "test-client-secret",
+            "quarkus.oidc.authentication.redirect-path": "/login/oauth2/code/oidc",
```
I'm not sure we need to set this. The [default](https://github.com/jhipster/generator-jhipster-quarkus/blob/main/generators/server/templates/quarkus/src/main/resources/application.properties.ejs#L90) is the same value.

```
quarkus.oidc.authentication.redirect-path=/login/oauth2/code/oidc
```

mraible commented on common/src/test/groovy/com/okta/cli/common/model/OidcPropertiesTest.groovy:
```diff
@@ -63,6 +67,95 @@ class OidcPropertiesTest {
+                "quarkus.oidc.client-id": "test-client-id",
+                "quarkus.oidc.credentials.secret": "test-client-secret",
+                "quarkus.oidc.authentication.redirect-path": "/login/oauth2/code/oidc",
+                "quarkus.oidc.application-type": "web-app"
```
The default is `hybrid`: https://github.com/jhipster/generator-jhipster-quarkus/blob/main/generators/server/templates/quarkus/src/main/resources/application.properties.ejs#L87

Are you changing this because Okta needs it or JHipster does. @danielpetisme Do you know what the difference is between `web-app`, `service`, and `hybrid` is when it comes to JHipster?

danielpetisme replied:
>Are you changing this because Okta needs it or JHipster does. @danielpetisme Do you know what the difference is between web-app, service, and hybrid is when it comes to JHipster?

I do! I wasted way too much hours on this...
- `service`:  targets system to system authentication (ie. microservices seeking for authentication, or in unit tests)
- `web-app`: targets a user dealing with a UI (like when you're actually using the JHipster web app)
- `hybrid`: well it's a mix of the 2 above
More info:
https://github.com/quarkusio/quarkus/blob/master/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java#L775

mraible replied:
In OAuth terms, it sounds like service is a resource server, a web app is for a UI that can redirect, and a hybrid is for both. JHipster's Spring Security support uses "hybrid" for its monoliths and gateways so you can log in AND they can be a resource server (for JHipster's Ionic and React Native modules). For microservices, it just sets up a resource server.

IMO, it doesn't seem like the Okta CLI should override this value. It should just default to what JHipster has.

danielpetisme commented:
Sorry guys I wasn't following the review...
You have plenty of relevant questions and most of them ring my bell but I must admit I didn't look at the code since the initial PR.

What's not clear for me it's the PR's scope. Are we dealing with only Quarkus ? only JHipster-Quarkus ? Both ?
From what I remember, my initial PR was covering only Quarkus. For me, JHipster QUarkus was mostly an extension of the vanilla Quarkus support. That's why for instance the `JHIPSTER_OIDC_LOGOUT_URL` was not initially generated (and I guess the same for the `application-type`).

mraible commented:
@danielpetisme It started with @bdemers adding Quarkus support based on your PR. Then I requested JHipster + Quarkus support. I wanted it because it's tedious to configure the groups, add your user to them, and create the `groups` claim in the ID token.

danielpetisme commented:
But everything is done in one big PR, right?
@bdemers would you like a review from my part too? I can do that if it makes sense to you.

bdemers commented:
@danielpetisme please, it's a bit fractured at the moment,

We have this PR #82, (mostly @danielpetisme's initial work, rebased)
Then on top of that, we have: #83 (adds micronaut support via OidcProperties)
And then on top of that is: mraible/okta-cli#1 (which refactors things to all for a more dynamic template, _mostly_ for jhipster)

I'll flatten this all out into a single separate PR in the morning (unless someone objects)

bdemers commented:
I just merged everything into this PR

mraible commented:
I grabbed the macOS binary from the bottom of [this PRs latest run](https://github.com/okta/okta-cli/actions/runs/559088192) to test things out. After clicking through all the permissions issues, here's my synopsis:

❌ `okta start spring-boot` 

Fails with NPE after it prints out the "set your password" URL:

```
Check your email
Verification code: [redacted]
New Okta Account created!
Your Okta Domain: https://dev-[redacted].okta.com
To set your password open this link:
https://dev-[redacted].okta.com/welcome/drp-uifYPuDcAV2EtA9t

java.lang.NullPointerException
	at com.okta.cli.common.service.DefaultSetupService.createOidcApplication(DefaultSetupService.java:178)
	at com.okta.cli.commands.Start.runCommand(Start.java:123)
	at com.okta.cli.commands.BaseCommand.call(BaseCommand.java:41)
	at com.okta.cli.commands.BaseCommand.call(BaseCommand.java:26)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1783)
	at picocli.CommandLine.access$900(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2150)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2144)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2108)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:1975)
	at picocli.CommandLine.execute(CommandLine.java:1904)
	at com.okta.cli.OktaCli.run(OktaCli.java:64)
	at com.okta.cli.OktaCli.main(OktaCli.java:54)
Execution time: 2 min. 8 s.
```

❌ `okta start angular` 

```
java.lang.NullPointerException
	at com.okta.cli.common.service.DefaultSetupService.createOidcApplication(DefaultSetupService.java:178)
	at com.okta.cli.commands.Start.runCommand(Start.java:123)
	at com.okta.cli.commands.BaseCommand.call(BaseCommand.java:41)
	at com.okta.cli.commands.BaseCommand.call(BaseCommand.java:26)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1783)
	at picocli.CommandLine.access$900(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2150)
```

✅ `okta apps create jhipster` (Spring Boot) 

❌ `okta apps create jhipster` (Quarkus)

Defaults to wrong redirect URI (should end with `/oidc` because it's a JHipster app).

```
Common defaults:
 JHipster - http://localhost:8080/login/oauth2/code/oidc
 Quarkus OIDC - http://localhost:8080/callback
 Spring Security - http://localhost:8080/login/oauth2/code/okta
Enter your Redirect URI(s) [http://localhost:8080/login/oauth2/code/okta]
```

If I manually input the correct redirect URI (`http://localhost:8080/login/oauth2/code/oidc`), it works. I didn't try accepting the default.

If I look in the created `.okta.env`, I don't think the application type is necessary since JHipster should control that.

```
export QUARKUS_OIDC_APPLICATION_TYPE="web-app"
```

Also, it'd be cool if the `JHIPSTER_OIDC_LOGOUT_URL` reused the `QUARKUS_OIDC_AUTH_SERVER_URL` variable in its value. I realize that's asking a lot. 😉

If I fail to run `source .okta.env`, the error message is kinda cryptic.

```
Feb 11, 2021 10:02:01 PM io.quarkus.runtime.ApplicationLifecycleManager run
ERROR: Failed to start application (with profile dev)
java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:779)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:330)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:334)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:702)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

This should be fixed on the JHipster side. It's failing because it can't connect to Keycloak.

✅ `okta apps create jhipster` (Micronaut)

The default values are correct, especially since it includes `8761` for the JHipster Registry. Should Micronaut be listed in the common defaults?

```
Common defaults:
 JHipster - http://localhost:8080/login/oauth2/code/oidc
 Quarkus OIDC - http://localhost:8080/callback
 Spring Security - http://localhost:8080/login/oauth2/code/okta
Enter your Redirect URI(s) [http://localhost:8080/oauth2/callback/oidc, http://localhost:8761/oauth2/callback/oidc]:
```

According to [Secure a Micronaut app with Okta](https://guides.micronaut.io/micronaut-oauth2-okta/guide/index.html), the Micronaut default should be `http://localhost:8080/oauth/callback/okta`.

After accepting the defaults, I realized that Micronaut might not provide an implementation of the JHipster Registry. Ideally, they should, but it could be a lot of work. They might just use the one that JHipster provides (written with Spring Boot). In that case, the `http://localhost:8761/logout` redirect should be `http://localhost:8761`. 

I'd be curious to see what @atomfrede thinks about this? Will the Micronaut blueprint provide its own registry like JHipster Registry?

What happens with Micronaut if I forget to `source .okta.env`? It fails with a LONG stack trace. It tells you the URL it's trying to connect to though.

```
22:12:08.755 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Bean definition [com.jhipster.demo.blog.config.LoggingConfiguration] could not be loaded: Error instantiating bean of type  [com.jhipster.demo.blog.security.JHipsterEndSessionEndpoint]

Message: Connect Error: Connection refused: localhost/127.0.0.1:9080
Path Taken: new JHipsterEndSessionEndpoint(EndSessionCallbackUrlBuilder endSessionCallbackUrlBuilder,OauthClientConfiguration clientConfiguration,[OpenIdProviderMetadata providerMetadata])
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [com.jhipster.demo.blog.config.LoggingConfiguration] could not be loaded: Error instantiating bean of type  [com.jhipster.demo.blog.security.JHipsterEndSessionEndpoint]
```

If I look at `.okta.env`, it has the following values that seem unnecessary.

```
export MICRONAUT_SECURITY_ENDPOINTS_LOGOUT_PATH="/logout"
export MICRONAUT_SECURITY_OAUTH2_CALLBACK_URI="/oauth2/callback/oidc"
```

These are the same values that are set by default in a JHipster + Micronaut app. Look in `src/main/resources/application.yml` after generating an app for proof.

### Summary

1. `okta start spring-boot` and `okta start angular` is broken after this PR. I didn't try other samples.
2. JHipster + Spring Boot works, but it doesn't include the `:8761` URL for the JHipster Registry.
3. JHipster + Quarkus defaults to the wrong URL and doesn't include the `:8761` redirect. Also, it overrides the application type, which it shouldn't when it's a JHipster + Quarkus app.
4. JHipster + Micronaut works but adds two environment variables to `.okta.env` that are unnecessary since their values are the same as the defaults. Also, `:8761/logout` is only necessary if Micronaut provides an implementation of the JHipster Registry. Otherwise, it should be `:8761` (with no context path).
5. Colors would be a fun addition. 🌈

bdemers commented:
Thanks @mraible!  I'll add some tests to catch that, improving the test coverage for the `start` command has been on my short list for a while

bdemers commented:
@mraible 

> Summary
> okta start spring-boot and okta start angular is broken after this PR. I didn't try other samples.

I broke this on the merge, fixed and added test

> JHipster + Spring Boot works, but it doesn't include the :8761 URL for the JHipster Registry.

Another merge issue, fixed and added tests

> JHipster + Quarkus defaults to the wrong URL and doesn't include the :8761 redirect. Also, it overrides the application type, which it shouldn't when it's a JHipster + Quarkus app.

Defaults to: `["http://localhost:8080/login/oauth2/code/oidc", "http://localhost:8761/login/oauth2/code/oidc"]` (added test)

@danielpetisme, is there any downside to NOT setting the app-type? (and assume the user has set this correctly), The quarkus app generator does set this value?

> JHipster + Micronaut works but adds two environment variables to .okta.env that are unnecessary since their values are the same as the defaults. Also, :8761/logout is only necessary if Micronaut provides an implementation of the JHipster Registry. Otherwise, it should be :8761 (with no context path).

These properties _can_ be changed via the prompt, but changing them could have side effects, see below 🤔 

> Colors would be a fun addition. 🌈

Yes! After this PR is merged ;)

> A thought after reading this: is it possible to suppress the redirect URI prompts for JHipster? It seems that if people change them, it'll cause things to fail. It might be confusing if you see JHipster and Quarkus defaults and don't know which to choose. > Then again, it might be handy for power users because they can set production URLs too.

This is tricky if you are creating a local test project, then there really is no reason to prompt, but... if you are doing anything other than localhost, we would need to prompt for the URL, (ngrok, heroku, etc).  And some of those cases we wouldn't want to set up the test port redirect URI.


## My 2-cents (assuming everything works):
Let's keep the redirect-uri prompt in the short term. Probably replacing it with a different set of prompts (more thought needed)

I'm all for removing any extra properties: e.g. `app-type`, just let me know.  That stuff is really easy to add tests for :)

mraible commented:
I gave it another go with [this build for macOS](https://github.com/okta/okta-cli/actions/runs/561928156):

✅ `okta register`
✅ `okta start spring-boot` 
✅ `okta start angular` 
✅ `okta apps create jhipster` (Spring Boot)
✅ `okta apps create jhipster` (Quarkus)
✅ `okta apps create jhipster` (Micronaut)

I still don't think Quarkus should override the application type for a JHipster app.

```
export QUARKUS_OIDC_APPLICATION_TYPE="web-app"
```

I commented out this line in `.okta.env` and everything still works.

The only thing I think this will affect is mobile apps. Ionic and React Native probably won't work unless the JHipster app is configured as a resource server too. That being said, I think it's OK for this release since we haven't tested React Native or Ionic with a Quarkus (or Micronaut) backend.

🚢 it!

mraible commented:
@bdemers I fixed the issue I mentioned above in https://github.com/okta/okta-cli/pull/85.

mraible commented:
If I try to reproduce the Ubuntu failure on my Ubuntu laptop, I get the following error:

```
Fatal error:com.oracle.svm.core.util.VMError$HostedError: Option name "HotSpotPostOnExceptions" has multiple definitions: org.graalvm.compiler.hotspot.meta.HotSpotExceptionDispatchPlugin$Options.HotSpotPostOnExceptions and org.graalvm.compiler.hotspot.meta.HotSpotNodePlugin$Options.HotSpotPostOnExceptions
	at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:68)
	at com.oracle.svm.hosted.option.HostedOptionParser.collectOptions(HostedOptionParser.java:89)
	at com.oracle.svm.hosted.option.HostedOptionParser.<init>(HostedOptionParser.java:62)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:258)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:529)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:119)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:561)
Error: Image build request failed with exit status 1
```

I'll keep trying.

mraible commented:
I was able to fix the issue above by downgrading GraalVM, from `21.0.0.2.r11-grl` to `20.3.1.2.r11-grl` (with SDKMAN). Since all builds pass, maybe we just need to restart CI?

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Okta CLI Tools 0.7.2-SNAPSHOT:
[INFO]
[INFO] Okta CLI Tools ..................................... SUCCESS [  1.233 s]
[INFO] Okta CLI Common .................................... SUCCESS [ 13.601 s]
[INFO] Okta CLI ........................................... SUCCESS [02:25 min]
[INFO] Okta CLI ITs ....................................... SUCCESS [02:53 min]
[INFO] Okta CLI Tools :: Test Coverage .................... SUCCESS [  0.270 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:34 min
[INFO] Finished at: 2021-03-02T13:08:51-07:00
[INFO] ------------------------------------------------------------------------
```

mraible commented:
I tested everything with the macOS binary from [this CI build](https://github.com/okta/okta-cli/actions/runs/614549933). Results:

✅ okta register
✅ okta start spring-boot
✅ okta start angular
✅ okta start react
✅ okta start vue

And JHipster tests:

✅ okta apps create jhipster (Spring Boot)
✅ okta apps create jhipster (Quarkus)
✅ okta apps create jhipster (Micronaut)

SHIP IT!! 🚀
